src/common/ad_mem.c

```diff
@@ -1678,10 +2024,23 @@ ad_reserve_addr(struct ad_blob *blob, int type, daos_size_t size,
 	tried = false;
 	while (1) {
 		if (arena == NULL) {
-			rc = arena_reserve(blob, type, NULL, &arena);
-			if (rc) {
-				D_ERROR("Failed to reserve new arena: %d\n", rc);
-				return 0;
+again:
+			node = d_binheap_remove_root(&blob->bb_arena_free_heap);
+			if (node == NULL) {
+				rc = arena_reserve(blob, type, NULL, &arena);
+				if (rc) {
+					D_ERROR("Failed to reserve new arena.\n");
+					return 0;
+				}
+			} else {
+				ad_node = container_of(node, struct ad_maxheap_node, mh_node);
+				rc = arena_load(blob, ad_node->mh_arena_id, &arena);
+				if (rc) {
+					D_DEBUG(DB_TRACE, "failed to load arena %u: %d\n",
+						ad_node->mh_arena_id, rc);
+					goto again;
```

Comment:
instead of using "goto",  "continue"  can do the same for you right? We also should check the returned value (I know I didn't but we should be very careful when it's in a loop).